server/src/main/java/org/elasticsearch/cluster/routing/allocation/Hosts.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.cluster.routing.allocation;
+
+import org.elasticsearch.cluster.routing.RoutingNode;
+import org.elasticsearch.common.Strings;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * {@link Hosts} hosts records the mapping relationship between nodes and hosts,
+ * which makes it easy to find nodes on the same host
+ */
+public class Hosts {
+    private final Map<String, Host> node2Host;
+
+    public Hosts(Iterable<RoutingNode> nodes) {
+        Map<String, HostBuilder> node2HostBuilder = new HashMap<>();
+        Map<String, HostBuilder> hostNameMap = new HashMap<>();
+        Map<String, HostBuilder> hostAddressMap = new HashMap<>();
+        for (RoutingNode checkNode : nodes) {
```

Comment:
> Assume you have one node with address only, one node with name only and one node with both. Depending on order, I think the node2Host map will contain either one or two hosts?
I also considered a similar problem, assuming we have three nodes. Their name and address are like 
- node1:name, address
- node2:none, address
- node3:name, none

If node1 is the last one to add., there will be two host in the map, otherwise just one. To solve this problem, we need to build an undirected graph and get the *number of connected components in an undirected graph*, each connected components is a host. 

As for original code, similar problem also exist. Node2 and node3 will be judged to have the same host as node1, but they(node2 and node3) will be judged to be not on the same host.

This scenario should not be common in a production environment.